Add Host to Org and Space Policy Automatically on Bind

micahlee opened the pull request:
This PR updates the Conjur service broker to include the Organization GUID and Space GUID from PCF into the host identity for the application, and to add the applications to Conjur `!layers` for the org and space. This allows a Conjur operator to permit an application to access resources based on the Space or Org in which it is deployed, rather than only on the application bind id itself.

Connected to #63 

Jenkins: https://jenkins.conjur.net/job/cyberark--conjur-service-broker/job/63-add-host-to-org-space-policy/

## Reviewer Notes

The acceptance criteria for this change are below, with references to the relevant PR changes:
- [ ] Flows are implemented to create host and add to space layer on bind
    > The flow changes to add the host to the space layer are in: https://github.com/cyberark/conjur-service-broker/pull/73/commits/122c41f3a6fdc7ebcbc2c2db546fd8c4c9f3677a

- [ ] Unit (RSpec) tests verify new org/space/app binding functionality
    > See changes in [bind_controller_spec.rb](https://github.com/cyberark/conjur-service-broker/blob/179f66f0fee2f45e5ff9ebaa89ac021ceb0237bb/spec/controllers/bind_controller_spec.rb#L57-L66)

- [ ] Unit (RSpec) tests verify backward compatibility when no org/space context is present 
    > See changes in [bind_controller_spec.rb](https://github.com/cyberark/conjur-service-broker/blob/179f66f0fee2f45e5ff9ebaa89ac021ceb0237bb/spec/controllers/bind_controller_spec.rb#L43-L55)

- [ ] Unit (RSpec) tests verify Conjur V4 functionality is maintained
    > See changes in [bind_controller_spec.rb](https://github.com/cyberark/conjur-service-broker/blob/179f66f0fee2f45e5ff9ebaa89ac021ceb0237bb/spec/controllers/bind_controller_spec.rb#L68-L83)

- [ ] Integration (Cucumber) tests verify new functionality
    > See commit: https://github.com/cyberark/conjur-service-broker/pull/73/commits/c8f1bfce72c71cc6d260753bf8c46e4cfbf37461

- [ ] README is updated with documentation on org / space layer functionality with recommendations for how to leverage the functionality / supported versions
    > See commit: https://github.com/cyberark/conjur-service-broker/pull/73/commits/0b6f86c55ce4486f0763e79109e2900180f00a70

- [ ] Issue created in [docs repo](https://github.com/cyberark/docs-cyberark-conjur-service-broker)  to update tile docs to describe new policy structure and bind behavior
    > https://github.com/cyberark/docs-cyberark-conjur-service-broker/issues/2

izgeri commented on README.md:
```diff
@@ -178,6 +178,62 @@ applications:
+  privileges: [ create, update, read ]
+  roles: !host service-broker
+
+- !permit
```
is this block a duplicate of the block above?

micahlee replied:
It's not, the one above is for the space `!policy`, this one is for the `!layer`.

As I'm thinking about it now, just the one for `!policy` may be sufficient. I will check that now.

micahlee replied:
No, I confirmed we need both the policy and layer permissions for the service broker. We are not able to check that the layer exists unless we have `read` access to it. We don't have that just from the policy permissions.

micahlee commented on CHANGELOG.md:
```diff
@@ -5,6 +5,9 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
 
 ## [Unreleased]
+### Changed
+- The service broker now adds application host's to a Conjur Layer for a Space when the
```
Yes, I'll fix that.

izgeri commented on lib/conjur_client.rb:
```diff
@@ -15,6 +19,10 @@ def v4_host_factory_id
     end
 
+    def v5?
+      ConjurClient.version == 5
```
I think the `ConjurClient` here is redundant

izgeri commented on spec/controllers/bind_controller_spec.rb:
```diff
@@ -1,7 +1,86 @@
+          allow(ENV).to receive(:[]).with('CONJUR_VERSION').and_return('4')
+        end
+
+        it "does ensure the policy structure exists" do
```
"does not ensure", right?

izgeri commented on app/models/service_binding.rb:
```diff
@@ -9,29 +9,31 @@ class HostNotFound < RuntimeError
   end
 
   class << self
-    def create(instance_id, binding_id, app_id)
```
out of curiosity, did you remove these just bc they're not being used? I wonder if it's worth leaving these if we might update to use the app GUID instead of the binding GUID at some point (I think that's why they are here now)

micahlee replied:
This is only my opinion, but I think it would be more clear to remove them since they aren't used now, and add the `app_guid` parameter back once we're ready to actually use it in a future changeset.

RafiS3 commented on README.md:
```diff
@@ -187,6 +242,12 @@ credentials in the application's environment, and might look something like
 `host/cf/0299a19d-7de4-4e98-89f6-372ac7c0521f` (for example, if your `CONJUR_POLICY`
 was set to `cf`).
```
What do you think about slightly changing the example of `cf` as the value of CONJUR_POLICY? If a company has multiple foundations, then `cf` is not a scalable name. Maybe we should recommend creating a policy called `cf` under `root` and then for each foundation, have its own policy like `production` then suggest that `CONJUR_POLICY` would have the value `cf/production`.
What do you think?

micahlee replied:
I think this is a great improvement! I will make the change.

micahlee commented:
@izgeri and @RafiS3, thank you both for the review notes! I believe I have updated the code or responded to all of the pending questions and comments.

Let me know if you have additional thoughts or ideas for improvement!

jtuttle commented on README.md:
```diff
@@ -178,14 +188,75 @@ applications:
+    # Layer to allow privileging an entire organzation to a resource
+    - !layer
+
+    # Policy for the Spaceß
```
Looks like a beta character snuck in on the end of this line somehow.

micahlee replied:
🤦‍♂️ Thanks!

With I accidentally hit option-s instead of cmd-s to save, that's what get's inserted. I will remove them characters.

jtuttle commented on README.md:
```diff
@@ -178,14 +188,75 @@ applications:
+    # Policy for the Spaceß
+    - !policy
+      # Space GUID from PCF.
+      # This may be obtained by running `cf space --guid {spaceß name}
```
There's a beta character here too.

jtuttle commented on features/support/world.rb:
```diff
@@ -46,16 +46,58 @@ def ci_secret_pass
+      member: !host pcf/#{host_id}
+    POLICY
+
+    puts policy
```
Was this just for debugging? Let's remove it if so.

jtuttle commented on features/support/world.rb:
```diff
@@ -46,16 +46,58 @@ def ci_secret_pass
-      POLICY
-      )
+      api.load_policy('root', policy)
       end
```
Mind fixing the indentation on this while you're at it?

jtuttle commented on spec/models/org_space_policy_spec.rb:
```diff
@@ -0,0 +1,49 @@
+    end
+
+    context "when org policy doesn't exist" do
+      before { allow_any_instance_of(Conjur::API).to receive(:resource).with('cucumber:policy:org_id').and_return(nonexistent_resource) }
```
Please add line breaks to the super long lines in this file (and the other specs). I generally use 80 chars.

micahlee commented:
@jtuttle and @RafiS3, thanks for the reviews! 

I believe I have addressed all of the feedback so far. Let me know if you think there is more than can be improved!

izgeri commented on README.md:
```diff
@@ -40,47 +40,57 @@ cf set-env conjur-service-broker SECURITY_USER_PASSWORD [value]
-- `CONJUR_ACCOUNT`: The account name for the Conjur instance you are connecting to.
-- `CONJUR_APPLIANCE_URL`: The URL of the Conjur appliance instance you are connecting to. If using high availability, this should be the URL for the Master host in the cluster. This is the URL that the service broker will use to communicate with Conjur.
+- `CONJUR_VERSION`: the version of your Conjur instance (`4` or `5`); defaults to 5.
+- `CONJUR_ACCOUNT`: the account name for the Conjur instance you are connecting to.
```
I wonder if we should put a note here about giving the URL of the master load balancer if using HA

izgeri commented on README.md:
```diff
@@ -40,47 +40,57 @@ cf set-env conjur-service-broker SECURITY_USER_PASSWORD [value]
+  >     - !policy prod
+  >       owner: !group cf-admin-group
+  >       body:
+  >       - !layer cf-apps
```
actually, from [here](https://github.com/cyberark/conjur-service-broker/blob/master/lib/conjur_client.rb#L14) it expects the layer to have fully qualified ID `CONJUR_POLICY/CONJUR_POLICY-apps`

so it should be `!layer cf/prod/cf/prod-apps` ... which is not very pretty. I am glad this is v4 only.

micahlee replied:
I updated this to reflect what I believe should be correct based on the link you provided.

It looks like this is for the host factory rather than the layer. And it doesn't have to be absolute in the policy file, but the ID does need to include the entire policy path.

Let me know if I'm missing something still.

izgeri replied:
This looks right. For consistency I prefer to advise that the hf layer have the same id, but since it's not technically required I leave it up to you

micahlee replied:
Sure thing, that makes a lot of sense!

izgeri commented on app/models/org_space_policy.rb:
```diff
@@ -0,0 +1,75 @@
+
+  class << self
+    def ensure_exists(org_id, space_id)
+      OrgSpacePolicy.new(org_id, space_id).ensure_exists
```
is it correct that until your provision changes get in, the bind request will fail because they won't have org / space policies set up?

micahlee replied:
It is true it will fail unless the policy is loaded before bind. This is documented in the README now as a manual step, and will be handled automatically by the service broker after the provision changes.

izgeri commented on README.md:
```diff
@@ -179,14 +189,75 @@ applications:
+  resource: !policy cbd7a05a-b304-42a9-8f66-6827ae6f78a1
+  privileges: [ create, update, read ]
+  # This should be the service broker identity you created with configuring the service broker
+  roles: !host service-broker
```
are these permits actually necessary? the service broker is a member of the group that owns the CONJUR_POLICY branch - I would expect that they inherit ownership of the space / org policy branches too.

micahlee replied:
Ah, I didn't have the pre-requisite policy set up correctly. Yes, when the ownership is set up as documented, these are not necessary. 

I will remove them. Thanks!

izgeri commented on app/models/org_space_policy.rb:
```diff
@@ -0,0 +1,75 @@
+  def ensure_exists
+    ensure_org_policy
+    ensure_space_policy
+    ensure_space_layer
```
any reason to not also `ensure_org_layer`?

micahlee replied:
The service broker, as coded, doesn't interact with the org layer directly on bind, so we don't need it to ensure it exist as a precondition.

That's the only reason I don't check it. If you would prefer it to also error if it's not present, I can do so.

izgeri replied:
We should at least have an automated test in your provision PR to ensure that the org layer exists and the host is (by inheritance) a member of it. If we are going to advise customers they can entitle the org policy layer, I'd like to have tests to ensure that will work.

Can you please add a note on this in the provision issue and ensure it gets implemented there?

micahlee replied:
Sure thing, will do!

micahlee commented:
Hey @izgeri, I have either updated the PR and marked comments resolved, or I've responded to all of the comments above. Let me know what you think!

izgeri commented on features/bind.feature:
```diff
@@ -1,4 +1,6 @@
 Feature: Binding
+  
+  @conjur-version-5
```
this doesn't seem great - we're effectively only testing bind now on v5, though we are still aiming to support v4. let's log an issue to address testing binding for v4 and v5 before we cut the next release

micahlee replied:
Thanks for calling that out, I made the change here but had forgotten to make a note to circle back to that. The issue for correcting this is here: https://github.com/cyberark/conjur-service-broker/issues/79
